lib/v2/wechat/data258.js

```diff
@@ -0,0 +1,138 @@
+const got = require('@/utils/got');
+const cheerio = require('cheerio');
+const { parseDate } = require('@/utils/parse-date');
+const timezone = require('@/utils/timezone');
+const { finishArticleItem } = require('@/utils/wechat-mp');
+const { RequestInProgressError } = require('@/errors');
+
+const parsePage = ($item, hyperlinkSelector, timeSelector) => {
+    const hyperlink = $item.find(hyperlinkSelector);
+    const title = hyperlink.text();
+    const link = hyperlink.attr('href');
+    const pubDate = timezone(parseDate($item.find(timeSelector).text(), 'YYYY-MM-DD HH:mm'), 8);
+    return {
+        title,
+        link,
+        pubDate,
+    };
+};
+
+const sleep = (time) => new Promise((resolve) => setTimeout(resolve, time * 1000));
+
+module.exports = async (ctx) => {
+    // !!! here we must use a lock to prevent other requests to break the anti-anti-crawler workarounds !!!
+    if ((await ctx.cache.get('data258:lock', false)) === '1') {
+        throw new RequestInProgressError('Another request is in progress, please try again later.');
+    }
+    // !!! here no need to acquire the lock, because the MP/category page has no crawler detection !!!
+
+    const id = ctx.params.id;
+
+    const limit = ctx.query.limit ? parseInt(ctx.query.limit) : 5;
+
+    const rootUrl = 'https://mp.data258.com';
+    const pageUrl = id ? `${rootUrl}/article/category/${id}` : rootUrl;
+
+    const response = await got(pageUrl);
+    const $ = cheerio.load(response.data);
+
+    const title = $('head title').text();
+    // title = title.endsWith('-微阅读') ? title.slice(0, title.length - 4) : title;
+    const description = $('meta[name="description"]').attr('content');
+
+    const categoryPage = $('ul.fly-list');
+
+    let items;
+    if (categoryPage && categoryPage.length) {
+        // got a category page
+        items = $(categoryPage)
+            .find('li')
+            .map((_, item) => parsePage($(item), 'h2 a', '.fly-list-info span'))
+            .get();
+    } else {
+        // got an MP page
+        items = $('ul.jie-row li')
+            .map((_, item) => parsePage($(item), 'a.jie-title', '.layui-hide-xs'))
+            .get();
+    }
+
+    items = items.slice(0, limit); // limit to avoid being anti-crawled
+
+    // !!! double-check !!!
+    if ((await ctx.cache.get('data258:lock', false)) === '1') {
+        throw new RequestInProgressError('Another request is in progress, please try again later.');
+    } else {
+        // !!! here we acquire the lock because the jump page has crawler detection !!!
+        await ctx.cache.set('data258:lock', '1', 60);
+    }
+
+    // !!! here we must use a for-loop to ensure the concurrency is 1 !!!
```

Comment:
You can use `tiny-async-pool`

Reply:
I consider there is no such need to use it if the concurrency is 1. Would you please describe the advantages of adopting it in this case? Making the flow body a function to improve the performance?

Reply:
No benefits. I was thinking this helps to tweak the concurrency(to 2) easily but after getting throttled multiple times I would consider a no for `tiny-async-pool`